### PR TITLE
Problem retest improvements

### PR DESCRIPTION
**Pull request by gogo4ds**
Improve request to the database for Problem Retest
Closes https://github.com/SoftUni-Internal/suls-issues/issues/3264

Implement some sort of confirmation before Retest is triggered
Closes https://github.com/SoftUni-Internal/suls-issues/issues/3263

### CODE REVIEW

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ProblemsController.cs [hunk 1]
```diff
@@ -540,37 +541,82 @@ public ActionResult DownloadAdditionalFiles(int id)
...
             {
-                this.Data.ParticipantScores.DeleteParticipantScores(id.Value);
+                this.Data.ParticipantScores.DeleteParticipantScores(model.Id);
                 this.Data.SaveChanges();
 
-                var submissions = this.Data.Submissions.All().Where(s => s.ProblemId == id).Select(s => s.Id).ToList();
-                submissions.ForEach(this.RetestSubmission);
+                try
+                {
+                    this.Data.Context.DbContext.Configuration.AutoDetectChangesEnabled = false;
```
Comment (jicata) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ProblemsController.cs]: Are there any significant performance gains with this approach? What about potential dangers?

File: Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ProblemsController.cs [hunk 2]
```diff
@@ -821,18 +867,14 @@ private void ValidateUploadedFile(string propertyName, HttpPostedFileBase file)
...
-                Processed = false,
-                Processing = false
-            };
-            this.Data.Context.Submissions.Attach(submission);
+            var submission = this.Data.Submissions.GetById(submissionId);
+
+            submission.Processed = false;
+
             var submissionEntry = this.Data.Context.Entry(submission);
             submissionEntry.Property(pr => pr.Processed).IsModified = true;
```
Comment (jicata) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ProblemsController.cs]: Have you considered using batch Update from the EntityFramework.Extended library? [More info here](https://github.com/zzzprojects/EntityFramework-Plus/wiki/EF-Batch-Update-%7C-Entity-Framework-Batch-Update-Objects-Without-Loading-Them)

File: Open Judge System/Data/OJS.Data/Repositories/SubmissionsForProcessingRepository.cs [hunk 3]
```diff
@@ -25,14 +27,49 @@ public void AddOrUpdate(int submissionId)
...
+                submissionForProcessing = new SubmissionForProcessing
                 {
                     SubmissionId = submissionId
                 };
+
                 this.Context.SubmissionsForProcessing.Add(submissionForProcessing);
             }
         }
 
+        public void BulkAddOrUpdate(IEnumerable<int> submissionIds)
```
Comment (jicata) [inline Open Judge System/Data/OJS.Data/Repositories/SubmissionsForProcessingRepository.cs]: This is not a "Bulk" operation per se. Simply overload the `AddOrUpdate` method instead.

File: Open Judge System/Data/OJS.Data/Repositories/SubmissionsForProcessingRepository.cs [hunk 4]
```diff
@@ -10,5 +14,71 @@ public SubmissionsForProcessingRepository(IOjsDbContext context)
...
+        }
+
+        public void AddOrUpdate(IEnumerable<int> submissionIds)
+        {
+            try
+            {
+                this.Context.DbContext.Configuration.AutoDetectChangesEnabled = false;
+                foreach (var submissionId in submissionIds)
+                {
+                    var submissionForProcessing = this.Context.SubmissionsForProcessing
```
Comment (jicata) [inline Open Judge System/Data/OJS.Data/Repositories/SubmissionsForProcessingRepository.cs]: Unify logic between this method and the single AddOrUpdate and reuse common functionality.

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Problems/RetestConfirmation.cshtml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+@using SharedResource = Resources.Areas.Contests.ContestsGeneral;
+
+@model OJS.Web.Areas.Administration.ViewModels.Problem.ProblemRetestViewModel
+
+@{
+    ViewBag.Title = Resource.Retest;
+}
+
+<h1 class="text-danger">&nbsp;<span class="glyphicon glyphicon-warning-sign"></span> @ViewBag.Title</h1>
+<p><span class="glyphicon glyphicon-circle-arrow-right"></span>&nbsp;&nbsp;<strong>@Model.ContestName</strong></p>
```
Comment (TsenkoTsenov) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Problems/RetestConfirmation.cshtml]: Are the `&nbsp;`'s on purpose?

Comment (gogo4ds) [reply Open Judge System/Web/OJS.Web/Areas/Administration/Views/Problems/RetestConfirmation.cshtml]: Yes

File: Open Judge System/OJS.Services/Data/SubmissionsForProcessing/SubmissionsForProcessingDataService.cs [hunk 6]
```diff
@@ -9,35 +10,39 @@
...
 
         public void AddOrUpdate(int submissionId)
         {
-            var submissionForProcessing = this.GetBySubmissionId(submissionId);
+            this.AddOrUpdateWithNoSaveChanges(submissionId);
+            this.submissionsForProcessing.SaveChanges();
+        }
 
-            if (submissionForProcessing != null)
+        //used optimization from this article https://msdn.microsoft.com/en-us/data/jj556205 for better performance
```
Comment (YanaSlavcheva) [inline Open Judge System/OJS.Services/Data/SubmissionsForProcessing/SubmissionsForProcessingDataService.cs]: 1/ space after `//`; 2/ Comment shall start with Uppercase letter

Comment (YanaSlavcheva) [reply Open Judge System/OJS.Services/Data/SubmissionsForProcessing/SubmissionsForProcessingDataService.cs]: Do we need this comment very much?

Comment (gogo4ds) [reply Open Judge System/OJS.Services/Data/SubmissionsForProcessing/SubmissionsForProcessingDataService.cs]: This comment was requested by @Innos and could prevent usage of `AutoDetectChangesEnabled` by someone else without reading about it.

File: Open Judge System/Web/OJS.Web/Areas/Administration/Views/Problems/RetestConfirmation.cshtml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+<div class="container ">
+    @using (Html.BeginForm("Retest", "Problems", new { area = "Administration" }))
```
Comment (rozay) [inline Open Judge System/Web/OJS.Web/Areas/Administration/Views/Problems/RetestConfirmation.cshtml]: Please, use `Mvc.Expressions`

**Pull request merged**